### PR TITLE
⚠️ Removes test related flags from binary

### LINKED ISSUE #1609: Binary exposes testing related flags
/kind bug

**What steps did you take and what happened:**
Currently the binary exposes the following flags:
```bash
cluster-api-provider-vsphere on  main [$?] via 🏎💨 v1.18.5
➜ ./binary --help
Usage of ./binary:
  -add_dir_header
    	If true, adds the file directory to the header of the log messages
  -alsologtostderr
    	log to standard error as well as files
  -credentials-file string
    	path to CAPV credentials file (default "/etc/capv/credentials.yaml")
  -enable-integration-tests
    	Enables integration tests
  -enable-keep-alive
    	DEPRECATED: feature to enable keep alive handler in vsphere sessions. This functionality is enabled by default now
  -enable-leader-election
    	Enable leader election for controller manager. Enabling this will ensure there is only one active controller manager. (default true)
  -enable-unit-tests
    	Enables unit tests (default true)
  -ginkgo.debug
    	If set, ginkgo will emit node output to files when running in parallel.
  -ginkgo.dryRun
    	If set, ginkgo will walk the test hierarchy without actually running anything.  Best paired with -v.
  -ginkgo.failFast
    	If set, ginkgo will stop running a test suite after a failure occurs.
  -ginkgo.failOnPending
    	If set, ginkgo will mark the test suite as failed if any specs are pending.
  -ginkgo.flakeAttempts int
    	Make up to this many attempts to run each spec. Please note that if any of the attempts succeed, the suite will not be failed. But any failures will still be recorded. (default 1)
  -ginkgo.focus value
    	If set, ginkgo will only run specs that match this regular expression. Can be specified multiple times, values are ORed.
  -ginkgo.noColor
    	If set, suppress color output in default reporter.
  -ginkgo.noisyPendings
    	If set, default reporter will shout about pending tests. (default true)
  -ginkgo.noisySkippings
    	If set, default reporter will shout about skipping tests. (default true)
  -ginkgo.parallel.node int
    	This worker nodes (one-indexed) node number.  For running specs in parallel. (default 1)
  -ginkgo.parallel.streamhost string
    	The address for the server that the running nodes should stream data to.
  -ginkgo.parallel.synchost string
    	The address for the server that will synchronize the running nodes.
  -ginkgo.parallel.total int
    	The total number of worker nodes.  For running specs in parallel. (default 1)
  -ginkgo.progress
    	If set, ginkgo will emit progress information as each spec runs to the GinkgoWriter.
  -ginkgo.randomizeAllSpecs
    	If set, ginkgo will randomize all specs together.  By default, ginkgo only randomizes the top level Describe, Context and When groups.
  -ginkgo.regexScansFilePath
    	If set, ginkgo regex matching also will look at the file path (code location).
  -ginkgo.reportFile string
    	Override the default reporter output file path.
  -ginkgo.reportPassed
    	If set, default reporter prints out captured output of passed tests.
  -ginkgo.seed int
    	The seed used to randomize the spec suite. (default 1662158123)
  -ginkgo.skip value
    	If set, ginkgo will only run specs that do not match this regular expression. Can be specified multiple times, values are ORed.
  -ginkgo.skipMeasurements
    	If set, ginkgo will skip any measurement specs.
  -ginkgo.slowSpecThreshold float
    	(in seconds) Specs that take longer to run than this threshold are flagged as slow by the default reporter. (default 5)
  -ginkgo.succinct
    	If set, default reporter prints out a very succinct report
  -ginkgo.trace
    	If set, default reporter prints out the full stack trace when a failure occurs
  -ginkgo.v
    	If set, default reporter print out all specs as they begin.
  -health-addr string
    	The address the health endpoint binds to. (default ":9440")
  -keep-alive-duration duration
    	idle time interval(minutes) in between send() requests in keepalive handler (default 5m0s)
  -kubeconfig string
    	Paths to a kubeconfig. Only required if out-of-cluster.
  -leader-election-id string
    	Name of the config map to use as the locking resource when configuring leader election. (default "capv-controller-manager-runtime")
  -log_backtrace_at value
    	when logging hits line file:N, emit a stack trace
  -log_dir string
    	If non-empty, write log files in this directory
  -log_file string
    	If non-empty, use this log file
  -log_file_max_size uint
    	Defines the maximum size a log file can grow to. Unit is megabytes. If the value is 0, the maximum file size is unlimited. (default 1800)
  -logtostderr
    	log to standard error instead of files (default true)
  -max-concurrent-reconciles int
    	The maximum number of allowed, concurrent reconciles. (default 10)
  -metrics-addr string
    	The address the metric endpoint binds to. (default "localhost:8080")
  -namespace string
    	Namespace that the controller watches to reconcile cluster-api objects. If unspecified, the controller watches for cluster-api objects across all namespaces.
  -network-provider string
    	network provider to be used by Supervisor based clusters.
  -one_output
    	If true, only write logs to their native severity level (vs also writing to each lower severity level)
  -pod-name string
    	The name of the pod running the controller manager. (default "capv-controller-manager")
  -profiler-address string
    	Bind address to expose the pprof profiler (e.g. localhost:6060)
  -root-dir string
    	Root project directory (default "../..")
  -skip_headers
    	If true, avoid header prefixes in the log messages
  -skip_log_headers
    	If true, avoid headers when opening log files
  -stderrthreshold value
    	logs at or above this threshold go to stderr (default 2)
  -sync-period duration
    	The interval at which cluster-api objects are synchronized (default 10m0s)
  -v value
    	number for the log level verbosity
  -vmodule value
    	comma-separated list of pattern=N settings for file-filtered logging
  -webhook-port int
    	Webhook Server port (set to 0 to disable)
```

From the output, it is obvious most of the `gingko` prefixed flags are coming in from the ginkgo dependency since it is being referred to via a non-test file.

A couple of other flags come from porting the legacy code from the CAPW codebase which exposed flags to choose between running and/or integration tests.

**What did you expect to happen:**
The `--help` section only shows flags relevant to the binary.

**Anything else you would like to add:**
This is not a breaking change since these flags are not used in the deployment files and in no way affect the functionality of the binary.

**Environment:**

- Cluster-api-provider-vsphere version: `v1.3.x`

### PR DESCRIPTION
**Pull request by srm09**
**What this PR does / why we need it**:
The binary was exposing flags used specifically by the tests which is misleading since the functionality of the binary is not altered by those flags.
Two main causes of the flags being exposed were:
1. ginkgo and gomega packages used in code required by the controllers
2. legacy code add flags to only run unit tests or integration tests.

This patch refactors the code to remove the flags from showing up on the binary.

**Which issue(s) this PR fixes**:
Fixes #1609

**Special notes for your reviewer**:
No changes to the image are made, since none of these flags were exposed in the manifests published by CAPV releases.

**Release note**:
```release-note
Clean up test flags from binary
```

### CODE REVIEW

**Comment by srm09** [comment]
Rebased off PR #1602 to get past linters and avoid last minute rebase errors
/hold wait for 1602 to get merged.

File: apis/v1beta1/vspheremachinetemplate_webhook.go [hunk 1]
```diff
@@ -17,57 +17,85 @@ limitations under the License.
...
 	}
 
-	return nil
+	req, err := admission.RequestFromContext(ctx)
+	if err != nil {
+		return apierrors.NewBadRequest(fmt.Sprintf("expected a admission.Request inside context: %v", err))
+	}
+
+	var allErrs field.ErrorList
+	if !topology.ShouldSkipImmutabilityChecks(req, newObj) &&
```
Comment (geetikabatra) [inline apis/v1beta1/vspheremachinetemplate_webhook.go:91]: Why are we adding this check?

File: controllers/serviceaccount_controller.go [hunk 2]
```diff
@@ -44,7 +44,6 @@ import (
 	"sigs.k8s.io/controller-runtime/pkg/source"
 
 	vmwarev1 "sigs.k8s.io/cluster-api-provider-vsphere/apis/vmware/v1beta1"
-	"sigs.k8s.io/cluster-api-provider-vsphere/pkg/builder"
```
[the remark this answers is not in the document]
Comment (geetikabatra) [reply controllers/serviceaccount_controller.go:47]: What exactly is this dependency doing?

File: apis/v1beta1/vspheremachinetemplate_webhook.go [hunk 3]
```diff
@@ -17,57 +17,85 @@ limitations under the License.
...
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
 )
 
-func (r *VSphereMachineTemplate) SetupWebhookWithManager(mgr ctrl.Manager) error {
+const machineTemplateImmutableMsg = "VSphereMachineTemplate spec.template.spec field is immutable. Please create a new resource instead."
+
+func (v *VSphereMachineTemplateWebhook) SetupWebhookWithManager(mgr ctrl.Manager) error {
 	return ctrl.NewWebhookManagedBy(mgr).
-		For(r).
+		For(&VSphereMachineTemplate{}).
```
[the remark this answers is not in the document]
Comment (srm09) [reply apis/v1beta1/vspheremachinetemplate_webhook.go:37]: This PR also has commits from #1602 as well, both these changes are required for CAPI bump to version v1.2. Here is the [relevant docs](https://cluster-api.sigs.k8s.io/developer/providers/v1.1-to-v1.2.html#required-api-changes-for-providers) excerpt from CAPI

**Comment by srm09** [comment]
> What exactly is this dependency doing?
This package was actually building a test context which was used for unit/integration testing of the controllers that we ported over for supervisor API types. This package was also being used by the `controllers` package directly, which was causing `ginkgo` flags to be added to the CAPV binary. Refactored the tests to remove that dependency.

File: controllers/serviceaccount_controller.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (srm09) [reply controllers/serviceaccount_controller.go:47]: This package was actually building a test context which was used for unit/integration testing of the controllers that we ported over for supervisor API types. This package was also being used by the controllers package directly, which was causing ginkgo flags to be added to the CAPV binary. Refactored the tests to remove that dependency.

**Comment by srm09** [comment]
/retitle ⚠️ Removes test related flags from binary

**Pull request merged**